openedx/core/djangoapps/session_inactivity_timeout/middleware.py

```diff
@@ -34,27 +39,63 @@ def process_request(self, request):
+            #   isn't being saved during that time, their last activity timestamp isn't being updated.
+            #   When they hit the inactivity timeout, it will be based on the last saved activity time.
+            #   So the effective timeout could be as short as: SESSION_INACTIVITY_TIMEOUT_IN_SECONDS - SESSION_ACTIVITY_SAVE_DELAY_SECONDS.
+            #   This means users might be logged out earlier than expected in some edge cases.
+            # .. setting_warning:  Must be smaller than SESSION_INACTIVITY_TIMEOUT_IN_SECONDS.
+            frequency_time_in_seconds = getattr(settings, "SESSION_ACTIVITY_SAVE_DELAY_SECONDS", 900)
 
             # Get the last time user made a request to server, which is stored in session data
-            last_touch = request.session.get(LAST_TOUCH_KEYNAME)
+            last_touch_str = request.session.get(LAST_TOUCH_KEYNAME)
 
             # have we stored a 'last visited' in session? NOTE: first time access after login
             # this key will not be present in the session data
-            if last_touch:
-                # compute the delta since last time user came to the server
-                time_since_last_activity = utc_now - last_touch
-
-                # did we exceed the timeout limit?
-                if time_since_last_activity > timedelta(seconds=timeout_in_seconds):
-                    # yes? Then log the user out
-                    del request.session[LAST_TOUCH_KEYNAME]
-                    auth.logout(request)
-                    return
-
-            request.session[LAST_TOUCH_KEYNAME] = utc_now
+            if last_touch_str:
+                try:
+                    last_touch = datetime.fromisoformat(last_touch_str)
+                    time_since_last_activity = current_time - last_touch
+
+                    has_exceeded_timeout_limit = time_since_last_activity > timedelta(seconds=timeout_in_seconds)
+                    if has_exceeded_timeout_limit:
+                        del request.session[LAST_TOUCH_KEYNAME]
+                        auth.logout(request)
+                        return
+                except (ValueError, TypeError) as e:
+                    # If parsing fails, log warning and then treat as if no timestamp exists
+                    log.warning("Parsing last touch time failed: %s", e)
+                    monitoring_utils.set_custom_attribute('session_inactivity.last_touch_status', 'last-touch-error')
+
+            else:
+                # .. custom_attribute_name: session_inactivity.last_touch_status
+                # .. custom_attribute_description: Tracks the status of session activity timestamps.
+                #   Values: 'first-login' (Tracks when users have no stored activity), 'last-touch-error' (failed to parse timestamp),
+                #   'last-touch-exceeded' (Marks when sessions are extended through the periodic save), 'last-touch-not-exceeded' (within save delay).
```

Comment:
Nit: This can be made a little more readable.
```suggestion
                #   Values: 'first-login' (Tracks when users have no stored activity),
                #     'last-touch-error' (failed to parse timestamp),
                #     'last-touch-exceeded' (Marks when sessions are extended through the periodic save),
                #     'last-touch-not-exceeded' (within save delay).
```

Reply:
updated

Reply:
@ttak-apphelix: I don't think you pushed your latest changes to github, so I was unable to review.